parachain/primitives/core/src/operating_mode.rs

```diff
@@ -0,0 +1,95 @@
+use codec::{Decode, Encode, FullCodec, MaxEncodedLen};
+use frame_support::{
+	dispatch::{DispatchResult, RawOrigin},
+	log, PalletError, RuntimeDebug, StorageValue,
+};
+use scale_info::TypeInfo;
+use sp_runtime::traits::BadOrigin;
+use sp_std::fmt::Debug;
+
+/// Error generated by the `OwnedBridgeModule` trait.
+#[derive(Encode, Decode, TypeInfo, PalletError)]
+pub enum BridgeModuleError {
+	/// All pallet operations are halted.
+	Halted,
+}
+
+/// Operating mode for a bridge module.
+pub trait OperatingMode: Send + Copy + Debug + FullCodec {
+	// Returns true if the bridge module is halted.
+	fn is_halted(&self) -> bool;
+}
+
+/// Basic operating modes for a bridges module (Normal/Halted).
+#[derive(
+	Encode,
+	Decode,
+	Clone,
+	Copy,
+	PartialEq,
+	Eq,
+	RuntimeDebug,
+	TypeInfo,
+	MaxEncodedLen,
+	serde::Serialize,
+	serde::Deserialize,
+)]
+pub enum BasicOperatingMode {
+	/// Normal mode, when all operations are allowed.
+	Normal,
+	/// The pallet is halted. All operations (except operating mode change) are prohibited.
+	Halted,
+}
+
+impl Default for BasicOperatingMode {
+	fn default() -> Self {
+		Self::Normal
+	}
+}
+
+impl OperatingMode for BasicOperatingMode {
+	fn is_halted(&self) -> bool {
+		*self == BasicOperatingMode::Halted
+	}
+}
+
+/// Bridge module with operating mode
+pub trait BridgeModule<T: frame_system::Config> {
+	/// The target that will be used when publishing logs related to this module.
+	const LOG_TARGET: &'static str;
+
+	type OperatingMode: OperatingMode;
+	type OperatingModeStorage: StorageValue<Self::OperatingMode, Query = Self::OperatingMode>;
+
+	/// Check if the module is halted.
+	fn is_halted() -> bool {
+		Self::OperatingModeStorage::get().is_halted()
+	}
+
+	/// Ensure that the origin is root.
+	fn ensure_root(origin: T::RuntimeOrigin) -> Result<(), BadOrigin> {
```

Comment:
I am not sure if this is possible but instead of duplicating the `ensure_root` function code here, we could add another type to the trait that implements `EnsureOrigin`.
```
type AllowedHaltOrigin: EnsureOrigin<Self::RuntimeOrigin>;
```

And then specialize it on pallets that implement the trait.
```
type AllowedHaltOrigin = EnsureRoot<AccountId>;
```

Reply:
https://github.com/Snowfork/snowbridge/pull/969/commits/aa4fb3d60d084e4cb603c169bdd8348f1e5d9c47